crates/prover-tracer/src/lib.rs

```
@@ -0,0 +1,283 @@
+            TracingOutput::Stderr => BoxMakeWriter::new(std::io::stderr),
+            // For OTLP, output traces also to stdout
+            TracingOutput::Otlp => BoxMakeWriter::new(std::io::stdout),
+            TracingOutput::File(path) => {
+                let appender = tracing_appender::rolling::never(".", path);
+                BoxMakeWriter::new(appender)
+            }
+        }
+    }
+}
+
+pub fn setup_tracing(config: &TracingConfig, version: &str) -> anyhow::Result<()> {
+    let writer = config.outputs.first().cloned().unwrap_or_default();
+
+    let mut layers = Vec::new();
+
+    // Setup instrumentation if both otlp agent url and
+    // otlp service name are provided as arguments
+    if config.outputs.contains(&TracingOutput::Otlp) {
+        if let (Some(otlp_agent), Some(otlp_service_name)) =
+            (&config.otlp_agent, &config.otlp_service_name)
+        {
+            let resources = build_resources(otlp_service_name, version);
+            let otlp_exporter = opentelemetry_otlp::SpanExporter::builder()
+                .with_tonic()
+                .with_endpoint(otlp_agent)
+                .build()?;
+
+            let batch_processor_config = BatchConfigBuilder::default()
+                .with_scheduled_delay(match std::env::var("OTLP_BATCH_SCHEDULED_DELAY") {
+                    Ok(v) => Duration::from_millis(v.parse::<u64>().unwrap_or(5_000)),
+                    _ => Duration::from_millis(5_000),
+                })
+                .with_max_queue_size(match std::env::var("OTLP_BATCH_MAX_QUEUE_SIZE") {
+                    Ok(v) => v.parse::<usize>().unwrap_or(2048),
+                    _ => 2048,
+                })
+                .with_max_export_batch_size(
+                    match std::env::var("OTLP_BATCH_MAX_EXPORTER_BATCH_SIZE") {
+                        Ok(v) => v.parse::<usize>().unwrap_or(512),
+                        _ => 512,
+                    },
+                );
```

Here too, the various `parse` calls could return an error rather than silently accept a malformed input and fabricate a value out of thin air. Same further down below this comment.